File: store/agents/openscad/skills/openscad/scripts/design.mjs
import {number,vector} from './expression.mjs';
const idPattern=/^[a-z][a-z0-9-]{0,31}$/;
const variablePattern=/^[a-zA-Z_][a-zA-Z0-9_]{0,47}$/;
function fields(value,allowed,label) {
  if (!value || typeof value!=='object' || Array.isArray(value)) throw new Error(label+' must be an object.');
  for (const key of Object.keys(value)) if (!allowed.includes(key)) throw new Error('Unknown '+label+' field: '+key);
}
function text(value,label,max=2000) {
  if (typeof value!=='string' || !value.trim() || value.length>max || /[\x00-\x08\x0b-\x1f]/.test(value)) throw new Error('Invalid '+label);
}
function list(value,label) {
  if (!Array.isArray(value) || value.length>30) throw new Error(label+' must be a list of at most 30 notes.');
  for (const item of value) text(item,label,1000);
}
function integer(value,min,max,label) {
  if (!Number.isInteger(value) || value<min || value>max) throw new Error(label+' must be an integer from '+min+' to '+max);
}
function ids(list,label) {
  if (!Array.isArray(list)) throw new Error(label+' must be a list.');
  const result=new Set();
  for (const item of list) {
    if (!item || !idPattern.test(item.id) || result.has(item.id)) throw new Error('Invalid or duplicate '+label+' ID.');
    result.add(item.id);
  }
  return result;
}
export function parametersFor(design,variant) {
  const parameters=Object.create(null);
  for (const [name,definition] of Object.entries(design.parameters)) {
    const value=Object.hasOwn(variant.values,name)?variant.values[name]:definition.value;
    number(value,{},name);
    if (typeof value!=='number' || value<definition.min || value>definition.max || definition.unit==='count'&&!Number.isInteger(value)) throw new Error(variant.id+': '+name+' is outside its saved limits.');
    parameters[name]=value;
  }
  for (const name of Object.keys(variant.values)) if (!Object.hasOwn(parameters,name)) throw new Error('Unknown variant parameter: '+name);
  return parameters;
}
export function regionFor(raw,parameters) {
  fields(raw,['shape','origin','size','radius','height','axis','repeat'],'region');
  const origin=vector(raw.origin,parameters,'Region origin');
  let region;
  if (raw.shape==='box') {
    if (['radius','height','axis'].some(key=>raw[key]!==undefined)) throw new Error('A box uses size, not cylinder fields.');
    region={shape:'box',origin,size:vector(raw.size,parameters,'Region size',{positive:true})};
  } else if (raw.shape==='cylinder') {
    if (raw.size!==undefined || !['x','y','z'].includes(raw.axis||'z')) throw new Error('Invalid cylinder region.');
    const radius=number(raw.radius,parameters,'Region radius'),height=number(raw.height,parameters,'Region height');
    if (radius<=0 || height<=0) throw new Error('Cylinder radius and height must be positive.');
    region={shape:'cylinder',origin,radius,height,axis:raw.axis||'z'};
  } else throw new Error('Region shape must be box or cylinder.');
  let copies=[origin];
  if (raw.repeat!==undefined) {
    if (!Array.isArray(raw.repeat) || raw.repeat.length>2) throw new Error('Use up to two repeat axes for a region.');
    for (const repeat of raw.repeat) {
      fields(repeat,['count','step'],'region repeat');
      const count=number(repeat.count,parameters,'Repeat count'),step=vector(repeat.step,parameters,'Repeat step');
      integer(count,1,32,'Region repeat count');
      if (copies.length*count>64) throw new Error('A region can contain at most 64 repeated probes.');
      copies=copies.flatMap(point=>Array.from({length:count},(_,i)=>point.map((value,axis)=>value+i*step[axis])));
    }
  }
  region.origins=copies; return region;
}
export function regionScad(region) {
  const literal=value=>JSON.stringify(value);
  const primitive=region.shape==='box'?'cube('+literal(region.size)+');':
    (region.axis==='x'?'rotate([0,90,0]) ':region.axis==='y'?'rotate([-90,0,0]) ':'')+'cylinder(r='+region.radius+',h='+region.height+',$fn=128);';
  return 'union() {\n'+region.origins.map(origin=>'translate('+literal(origin)+') '+primitive).join('\n')+'\n}';
}
export function checkFor(check,parameters) {
  const result={...check};
  if (check.kind==='bounds') {
    result.size=vector(check.size,parameters,'Required size',{positive:true});
    result.origin=vector(check.origin,parameters,'Required origin');
    result.tolerance=number(check.tolerance??.02,parameters,'Bounds tolerance');
    if (result.tolerance<=0 || result.tolerance>1) throw new Error('Bounds tolerance must be greater than zero and no more than 1 mm.');
  }
  if (check.kind==='volume') {
    if (!Array.isArray(check.range) || check.range.length!==2) throw new Error('Volume range needs minimum and maximum.');
    result.range=check.range.map(value=>number(value,parameters,'Volume limit'));
    if (result.range[0]<=0 || result.range[1]<result.range[0]) throw new Error('Invalid volume range.');
  }
  if (['empty','contains'].includes(check.kind)) result.region=regionFor(check.region,parameters);
  if (check.kind==='separated') {
    result.translation=vector(check.translation||[0,0,0],parameters,'Second part translation');
    result.rotation=vector(check.rotation||[0,0,0],parameters,'Second part rotation');
  }
  return result;
}
export function validateDesign(input) {
  const design=structuredClone(input);
  fields(design,['spec','title','brief','assumptions','assembly','sourceFiles','parameters','variants','parts','checks','process','preview'],'design');
  if (design.spec!==1) throw new Error('design.json requires spec: 1.');
  text(design.title,'title',100);text(design.brief,'brief',4000);
  list(design.assumptions,'Assumptions');list(design.assembly,'Assembly/use instructions');
  if (!Array.isArray(design.sourceFiles) || !design.sourceFiles.includes('model.scad') || !design.sourceFiles.includes('design.json') || design.sourceFiles.length>32) throw new Error('sourceFiles must include model.scad and design.json, with at most 32 files.');
  const names=new Set();
  for (const name of design.sourceFiles) {
    if (typeof name!=='string' || name.length>180 || !/^[a-zA-Z0-9][a-zA-Z0-9._/-]*$/.test(name) || name.split('/').some(component=>!component || component.startsWith('.')) || !/\.(scad|json|stl|svg|dxf|txt|md)$/i.test(name) || /^(handoff|rebuild)(\/|$)/i.test(name) || /^(part\.stl|preview\.scad|rebuild\.md)$/i.test(name)) throw new Error('Unsafe or reserved source path: '+name);
    if (names.has(name.toLowerCase())) throw new Error('Duplicate source path: '+name);
    names.add(name.toLowerCase());
  }
  if (!design.parameters || typeof design.parameters!=='object' || Array.isArray(design.parameters) || !Object.keys(design.parameters).length || Object.keys(design.parameters).length>64) throw new Error('Use 1–64 named numeric parameters.');
  for (const [name,definition] of Object.entries(design.parameters)) {
    if (!variablePattern.test(name) || name==='id' || name.startsWith('harness_')) throw new Error('Invalid or reserved parameter name: '+name);
    fields(definition,['value','min','max','unit','label'],'parameter '+name);
    for (const field of ['value','min','max']) if (typeof definition[field]!=='number') throw new Error(name+' '+field+' must be a number.');
    number(definition.min);number(definition.max);number(definition.value);
    if (definition.unit==='count'&&!Number.isInteger(definition.value)) throw new Error(name+' default count must be an integer.');
    if (definition.min>definition.max || definition.value<definition.min || definition.value>definition.max || !['mm','degrees','count','ratio'].includes(definition.unit)) throw new Error('Invalid limits, default or unit for '+name);
    if (definition.label!==undefined) text(definition.label,'parameter label',100);
  }
  const variants=ids(design.variants,'variant'),parts=ids(design.parts,'part');
  integer(variants.size,1,4,'Variants');integer(parts.size,1,8,'Parts');
  const exportNames=new Set();
  for (const variant of variants) for (const part of parts) {
    const name=variant+'-'+part;
    if (exportNames.has(name)) throw new Error('Ambiguous variant/part export name: '+name+'. Choose distinct IDs.');
    exportNames.add(name);
  }
  for (const variant of design.variants) {
    fields(variant,['id','label','values','notes'],'variant');
    text(variant.label,'variant label',100);list(variant.notes,'Variant notes');
    if (!variant.values || typeof variant.values!=='object' || Array.isArray(variant.values)) throw new Error('Variant values must be an object.');
  }
  for (const part of design.parts) {
    fields(part,['id','label','quantity','printRotation','shells','views','notes'],'part');
    text(part.label,'part label',100);list(part.notes,'Part notes');
    integer(part.quantity,1,100,'Part quantity');integer(part.shells,1,16,'Expected connected surface shells');
    part.printRotation=vector(part.printRotation,{},'Print rotation');
    fields(part.views,['top','front','side'],'part views');
    for (const name of ['top','front','side']) {
      fields(part.views[name],['at'],'view '+name);
      // at=null is an external silhouette; a numeric expression is a section.
      if (part.views[name].at===undefined) throw new Error('Each view declares at (section plane), or null for a silhouette.');
    }
  }
  fields(design.preview,['variant','part'],'preview');
  if (!variants.has(design.preview.variant)||!parts.has(design.preview.part)) throw new Error('Preview must name an exported variant and part.');
  fields(design.process,['material','bedMM','edgeMarginMM','nozzleMM','notes'],'process');
  text(design.process.material,'material',100);list(design.process.notes,'Process notes');
  design.process.bedMM=vector(design.process.bedMM,{},'Printer build envelope',{positive:true});
  design.process.edgeMarginMM=number(design.process.edgeMarginMM);design.process.nozzleMM=number(design.process.nozzleMM);
  if (design.process.edgeMarginMM<0 || design.process.nozzleMM<=0) throw new Error('Review edge margin and nozzle diameter.');
  const checks=ids(design.checks,'check');integer(checks.size,1,32,'Checks');
  const checkNames=new Set();
  for (const variant of variants) {
    for (const name of [...checks,...[...parts].flatMap(part=>['topology','reopen','bed'].map(suffix=>part+'-'+suffix))]) {
      const id=variant+'-'+name;
      if (checkNames.has(id)) throw new Error('Duplicate or reserved generated check ID: '+id);
      checkNames.add(id);
    }
  }
  const permitted={
    bounds:['size','origin','tolerance'],volume:['range'],empty:['region'],contains:['region'],separated:['other','translation','rotation']
  };
  for (const check of design.checks) {
    if (!Object.hasOwn(permitted,check.kind)) throw new Error('Unknown check kind: '+check.kind);
    fields(check,['id','label','kind','part',...permitted[check.kind]],'check '+check.id);
    text(check.label,'check label',200);
    if (!parts.has(check.part) || check.kind==='separated' && (!parts.has(check.other) || check.other===check.part)) throw new Error('Check names an unknown or identical second part.');
  }
  for (const part of design.parts) if (!design.checks.some(check=>check.part===part.id && check.kind==='bounds')) throw new Error('Every part needs a saved bounds requirement: '+part.id);
  for (const variant of design.variants) {
    const parameters=parametersFor(design,variant);
    for (const part of design.parts) for (const view of Object.values(part.views)) if (view.at!==null) number(view.at,parameters,'Section position');
    for (const check of design.checks) checkFor(check,parameters);
  }
  return design;
}

File: store/agents/freecad/skills/freecad/scripts/design.py
"""The saved job contract. Standard-library only so validation needs no CAD runtime."""
import math
import os
import re


def fail(message):
    raise ValueError("design.json: " + message)


def text(value, name, limit=2000):
    if not isinstance(value, str) or not value.strip() or len(value) > limit:
        fail(name + " must be nonempty text (at most %d characters)" % limit)
    return value


def number(value, name, minimum=None, maximum=1000000):
    if isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(value):
        fail(name + " must be a finite number")
    if minimum is not None and value < minimum or maximum is not None and value > maximum:
        fail(name + " is outside the supported range")
    return value


def vector(value, name, positive=False):
    if not isinstance(value, list) or len(value) != 3:
        fail(name + " must contain three millimetre coordinates")
    for component in value:
        number(component, name, 0.000001 if positive else -1000000)
    return value


def keys(value, required, optional, name):
    if not isinstance(value, dict):
        fail(name + " must be an object")
    missing = set(required) - set(value)
    unknown = set(value) - set(required) - set(optional)
    if missing:
        fail(name + " is missing " + ", ".join(sorted(missing)))
    if unknown:
        fail(name + " has unknown fields: " + ", ".join(sorted(unknown)))


def source_path(value):
    text(value, "sourceFiles entry", 240)
    components = value.split("/")
    if (os.path.isabs(value) or "\\" in value or ":" in value or
            any(not c or c.startswith(".") for c in components) or
            any(ord(c) < 32 for c in value)):
        fail("sourceFiles must be visible, relative paths inside the project")
    if components[0].lower() in {"handoff", "tools", "node_modules", "part.step", "part.stl", "part.fcstd", "candidate.step", "rebuild.txt"}:
        fail("sourceFiles cannot include generated outputs or tools: " + value)
    return value


def validate(value):
    keys(value, ["spec", "title", "brief", "units", "parts", "checks"],
         ["references", "sourceFiles", "assumptions", "assembly", "hardware"], "project")
    if value["spec"] != 1 or isinstance(value["spec"], bool):
        fail("spec must be 1")
    if value["units"] != "mm":
        fail("units must be mm; convert the user's measurements explicitly")
    text(value["title"], "title", 160)
    text(value["brief"], "brief", 6000)
    parts = value["parts"]
    if not isinstance(parts, list) or not 1 <= len(parts) <= 32:
        fail("parts must contain 1–32 manufactured parts")
    objects, ids = set(), set()
    for part in parts:
        keys(part, ["id", "object", "label", "material"], ["quantity", "notes"], "part")
        for field in ["id", "object"]:
            if not isinstance(part[field], str) or not re.fullmatch(r"[A-Za-z][A-Za-z0-9_-]{0,63}", part[field]):
                fail("part " + field + " must be a short ASCII identifier")
        if part["id"].lower() in ids or part["object"] in objects:
            fail("part IDs (case-insensitive) and FreeCAD object names must be unique")
        ids.add(part["id"].lower())
        objects.add(part["object"])
        text(part["label"], "part label", 160)
        text(part["material"], "part material", 240)
        quantity = part.get("quantity", 1)
        if type(quantity) is not int or not 1 <= quantity <= 1000:
            fail("part quantity must be an integer from 1 to 1000")
        if "notes" in part:
            text(part["notes"], "part notes")
    references = value.get("references", [])
    if not isinstance(references, list) or len(references) > 64:
        fail("references must be a list of up to 64 FreeCAD object names")
    for name in references:
        if not isinstance(name, str) or not re.fullmatch(r"[A-Za-z][A-Za-z0-9_-]{0,63}", name) or name in objects:
            fail("reference object names must be unique, distinct from manufactured parts")
        objects.add(name)
    sources = value.get("sourceFiles", ["part.FCMacro", "design.json"])
    if not isinstance(sources, list) or not 1 <= len(sources) <= 128:
        fail("sourceFiles must contain 1–128 explicitly selected source files")
    for path in sources:
        source_path(path)
    if len(set(path.lower() for path in sources)) != len(sources):
        fail("sourceFiles cannot contain duplicate names")
    if not {"part.FCMacro", "design.json"}.issubset(sources):
        fail("sourceFiles must include part.FCMacro and design.json")
    for field in ["assumptions", "assembly", "hardware"]:
        if field in value:
            if not isinstance(value[field], list) or len(value[field]) > 100:
                fail(field + " must contain at most 100 text entries")
            for entry in value[field]:
                text(entry, field + " entry")
    checks = value["checks"]
    if not isinstance(checks, list) or not 1 <= len(checks) <= 256:
        fail("checks must contain 1–256 measurable requirements")
    check_ids = set()
    for check in checks:
        if not isinstance(check, dict):
            fail("each check must be an object")
        kind = check.get("type")
        schemas = {
            "bounds": (["part", "size"], ["origin", "tolerance"]),
            "clearance": (["parts", "minimum"], ["maximum", "tolerance"]),
            "no-overlap": (["parts"], ["toleranceVolume"]),
            "contains": (["part", "other"], ["toleranceVolume"]),
            "bore": (["part", "origin", "axis", "diameter", "length", "wall"], ["tolerance"]),
        }
        if not isinstance(kind, str) or kind not in schemas:
            fail("unsupported check type: " + str(kind))
        required, optional = schemas[kind]
        keys(check, ["id", "type", "reason"] + required, optional, "check")
        text(check["id"], "check id", 100)
        text(check["reason"], "check reason", 1000)
        if check["id"] in check_ids:
            fail("check IDs must be unique")
        check_ids.add(check["id"])
        names = [check[field] for field in ["part", "other"] if field in check]
        if "parts" in check:
            if not isinstance(check["parts"], list) or len(check["parts"]) != 2:
                fail("check parts must name exactly two different objects")
            names += check["parts"]
        if any(not isinstance(name, str) or name not in objects for name in names) or len(set(names)) != len(names):
            fail("checks must name distinct objects declared in parts or references")
        for field in ["origin", "axis", "size"]:
            if field in check:
                vector(check[field], field, field == "size")
        for field in ["minimum", "maximum", "tolerance", "toleranceVolume"]:
            if field in check:
                number(check[field], field, 0)
        for field in ["diameter", "length", "wall"]:
            if field in check:
                number(check[field], field, 0.000001)
        if kind == "clearance" and check.get("maximum", float("inf")) < check["minimum"]:
            fail("clearance maximum cannot be less than minimum")
        if kind == "bore":
            if sum(component ** 2 for component in check["axis"]) < 1e-12:
                fail("bore axis cannot be zero")
            tolerance = check.get("tolerance", 0.02)
            if tolerance >= min(check["diameter"] / 2, check["length"] / 2, check["wall"]):
                fail("bore tolerance must be smaller than its radius, half-depth and wall")
    return value
